pkg/ottl/ottlfuncs/func_parse_key_value.go

```diff
@@ -0,0 +1,88 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package ottlfuncs
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal/parseutils"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+)
+
+type ParseKeyValueArguments[K any] struct {
+	Target        ottl.StringGetter[K]
+	Delimiter     ottl.Optional[string]
+	PairDelimiter ottl.Optional[string]
+}
+
+func NewParseKeyValueFactory[K any]() ottl.Factory[K] {
+	return ottl.NewFactory("ParseKeyValue", &ParseKeyValueArguments[K]{}, createParseKeyValueFunction[K])
+}
+
+func createParseKeyValueFunction[K any](_ ottl.FunctionContext, oArgs ottl.Arguments) (ottl.ExprFunc[K], error) {
+	args, ok := oArgs.(*ParseKeyValueArguments[K])
+
+	if !ok {
+		return nil, fmt.Errorf("ParseKeyValueFactory args must be of type *ParseKeyValueArguments[K]")
+	}
+
+	return parseKeyValue[K](args.Target, args.Delimiter, args.PairDelimiter)
+}
+
+func parseKeyValue[K any](target ottl.StringGetter[K], d ottl.Optional[string], p ottl.Optional[string]) (ottl.ExprFunc[K], error) {
+	delimiter := "="
+	if !d.IsEmpty() {
+		if d.Get() == "" {
+			return nil, fmt.Errorf("delimiter cannot be set to an empty string")
+		}
+		delimiter = d.Get()
+	}
+
+	pairDelimiter := " "
+	if !p.IsEmpty() {
+		if p.Get() == "" {
+			return nil, fmt.Errorf("pair delimiter cannot be set to an empty string")
+		}
+		pairDelimiter = p.Get()
+	}
+
+	if pairDelimiter == delimiter {
+		return nil, fmt.Errorf("pair delimiter %q cannot be equal to delimiter %q", pairDelimiter, delimiter)
+	}
+
+	return func(ctx context.Context, tCtx K) (any, error) {
+		source, err := target.Get(ctx, tCtx)
+		if err != nil {
+			return nil, err
+		}
+
+		if source == "" {
+			return nil, fmt.Errorf("cannot parse from empty target")
```

Comment:
Yea that's a good point. It can definitely be changed to just a trivial case, but I do think there's value in getting some kind of feedback about the target if it is empty to aid in configuring the function. I'm not sure though, maybe that's not really a big point of concern, in which case making it trivial seems fine to me.

Reply:
[stanza throws an error so we should too.](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/6f1b1d2a5c2a690e138feac4712c7af01b65f2a4/pkg/stanza/operator/parser/keyvalue/keyvalue.go#L97-L99)

Reply:
Thanks for calling that out. I just wanted to explore that option, but I think we're good here.